File: common/c-api/table.cpp
#include <string>
#include <vector>

#include "../dbconnector.h"
#include "../table.h"
#include "table.h"
#include "util.h"

using namespace swss;
using namespace std;

SWSSResult SWSSTable_new(SWSSDBConnector db, const char *tableName, SWSSTable *outTbl) {
    SWSSTry(*outTbl = (SWSSTable) new Table((DBConnector *)db, string(tableName)));
}

SWSSResult SWSSTable_free(SWSSTable tbl) {
    SWSSTry(delete (Table *)tbl);
}

SWSSResult SWSSTable_get(SWSSTable tbl, const char *key, SWSSFieldValueArray *outValues,
                         int8_t *outExists) {
    SWSSTry({
        vector<FieldValueTuple> fvs;
        bool exists = ((Table *)tbl)->get(string(key), fvs);
        if (exists) {
            *outValues = makeFieldValueArray(fvs);
            *outExists = 1;
        } else {
            *outExists = 0;
        }
    });
}

SWSSResult SWSSTable_hget(SWSSTable tbl, const char *key, const char *field, SWSSString *outValue,
                          int8_t *outExists) {
    SWSSTry({
        string s;
        bool exists = ((Table *)tbl)->hget(string(key), string(field), s);
        if (exists) {
            *outValue = makeString(move(s));
            *outExists = 1;
        } else {
            *outExists = 0;
        }
    });
}

SWSSResult SWSSTable_set(SWSSTable tbl, const char *key, SWSSFieldValueArray values) {
    SWSSTry({
        vector<FieldValueTuple> fvs = takeFieldValueArray(values);
        ((Table *)tbl)->set(string(key), fvs);
    });
}

SWSSResult SWSSTable_hset(SWSSTable tbl, const char *key, const char *field, SWSSStrRef value) {
    SWSSTry(((Table *)tbl)->hset(string(key), string(field), takeStrRef(value)));
}

SWSSResult SWSSTable_del(SWSSTable tbl, const char *key) {
    SWSSTry(((Table *)tbl)->del(string(key)));
}

SWSSResult SWSSTable_hdel(SWSSTable tbl, const char *key, const char *field) {
    SWSSTry(((Table *)tbl)->hdel(string(key), string(field)));
}

SWSSResult SWSSTable_getKeys(SWSSTable tbl, SWSSStringArray *outKeys) {
    SWSSTry({
        vector<string> keys;
        ((Table *)tbl)->getKeys(keys);
        *outKeys = makeStringArray(move(keys));
    });
}

File: crates/swss-common/src/types/table.rs
use super::*;
use crate::bindings::*;
use std::ptr;

/// Rust wrapper around `swss::Table`.
#[derive(Debug)]
pub struct Table {
    name: String,
    ptr: SWSSTable,
    _db: DbConnector,
}

impl Table {
    pub fn new(db: DbConnector, table_name: &str) -> Result<Self> {
        let table_name_copy = table_name.to_string();
        let table_name = cstr(table_name);
        let ptr = unsafe { swss_try!(p_tbl => SWSSTable_new(db.ptr, table_name.as_ptr(), p_tbl))? };
        Ok(Self {
            name: table_name_copy,
            ptr,
            _db: db,
        })
    }

    pub fn get(&self, key: &str) -> Result<Option<FieldValues>> {
        let key = cstr(key);
        let mut arr = SWSSFieldValueArray {
            len: 0,
            data: ptr::null_mut(),
        };
        let exists = unsafe { swss_try!(p_exists => SWSSTable_get(self.ptr, key.as_ptr(), &mut arr, p_exists))? };
        let maybe_fvs = if exists == 1 {
            Some(unsafe { take_field_value_array(arr) })
        } else {
            None
        };
        Ok(maybe_fvs)
    }

    pub fn hget(&self, key: &str, field: &str) -> Result<Option<CxxString>> {
        let key = cstr(key);
        let field = cstr(field);
        let mut val: SWSSString = ptr::null_mut();
        let exists = unsafe {
            swss_try!(p_exists => SWSSTable_hget(self.ptr, key.as_ptr(), field.as_ptr(), &mut val, p_exists))?
        };
        let maybe_fvs = if exists == 1 {
            Some(unsafe { CxxString::take(val).unwrap() })
        } else {
            None
        };
        Ok(maybe_fvs)
    }

    pub fn set<I, F, V>(&self, key: &str, fvs: I) -> Result<()>
    where
        I: IntoIterator<Item = (F, V)>,
        F: AsRef<[u8]>,
        V: Into<CxxString>,
    {
        let key = cstr(key);
        let (arr, _k) = make_field_value_array(fvs);
        unsafe { swss_try!(SWSSTable_set(self.ptr, key.as_ptr(), arr)) }
    }

    pub fn hset(&self, key: &str, field: &str, val: &CxxStr) -> Result<()> {
        let key = cstr(key);
        let field = cstr(field);
        unsafe { swss_try!(SWSSTable_hset(self.ptr, key.as_ptr(), field.as_ptr(), val.as_raw())) }
    }

    pub fn del(&self, key: &str) -> Result<()> {
        let key = cstr(key);
        unsafe { swss_try!(SWSSTable_del(self.ptr, key.as_ptr())) }
    }

    pub fn hdel(&self, key: &str, field: &str) -> Result<()> {
        let key = cstr(key);
        let field = cstr(field);
        unsafe { swss_try!(SWSSTable_hdel(self.ptr, key.as_ptr(), field.as_ptr())) }
    }

    pub fn get_keys(&self) -> Result<Vec<String>> {
        unsafe {
            let arr = swss_try!(p_arr => SWSSTable_getKeys(self.ptr, p_arr))?;
            Ok(take_string_array(arr))
        }
    }

    pub fn get_name(&self) -> &str {
        &self.name
    }
}

impl Drop for Table {
    fn drop(&mut self) {
        unsafe { swss_try!(SWSSTable_free(self.ptr)).expect("Dropping Table") };
    }
}

unsafe impl Send for Table {}

#[cfg(feature = "async")]
impl Table {
    async_util::impl_basic_async_method!(new_async <= new(db: DbConnector, table_name: &str) -> Result<Self>);
    async_util::impl_basic_async_method!(get_async <= get(&self, key: &str) -> Result<Option<FieldValues>>);
    async_util::impl_basic_async_method!(hget_async <= hget(&self, key: &str, field: &str) -> Result<Option<CxxString>>);
    async_util::impl_basic_async_method!(
        set_async <= set<I, F, V>(&self, key: &str, fvs: I) -> Result<()>
                     where
                         I: IntoIterator<Item = (F, V)> + Send,
                         F: AsRef<[u8]>,
                         V: Into<CxxString>,
    );
    async_util::impl_basic_async_method!(hset_async <= hset(&self, key: &str, field: &str, value: &CxxStr) -> Result<()>);
    async_util::impl_basic_async_method!(del_async <= del(&self, key: &str) -> Result<()>);
    async_util::impl_basic_async_method!(hdel_async <= hdel(&self, key: &str, field: &str) -> Result<()>);
    async_util::impl_basic_async_method!(get_keys_async <= get_keys(&self) -> Result<Vec<String>>);
}
